sdk/src/wallet/core/builder.rs

```diff
@@ -178,46 +175,95 @@ where
             self.secret_manager = secret_manager;
         }
 
-        let restored_bip_path = loaded_wallet_builder.as_ref().and_then(|builder| builder.bip_path);
+        let loaded_address = loaded_wallet_builder
+            .as_ref()
+            .and_then(|builder| builder.address.clone());
+
+        // May use a previously stored address it wasn't provided
+        if let Some(address) = &self.address {
+            if let Some(loaded_address) = &loaded_address {
+                if address != loaded_address {
+                    return Err(crate::wallet::Error::WalletAddressMismatch(address.clone()));
+                }
+            }
+        } else {
+            self.address = loaded_address;
+        }
+
+        let loaded_bip_path = loaded_wallet_builder.as_ref().and_then(|builder| builder.bip_path);
 
         // May use a previously stored BIP path if it wasn't provided
         if let Some(bip_path) = self.bip_path {
-            if let Some(restored_bip_path) = restored_bip_path {
-                if bip_path != restored_bip_path {
+            if let Some(loaded_bip_path) = loaded_bip_path {
+                if bip_path != loaded_bip_path {
                     return Err(crate::wallet::Error::BipPathMismatch {
                         new_bip_path: Some(bip_path),
-                        old_bip_path: Some(restored_bip_path),
+                        old_bip_path: Some(loaded_bip_path),
                     });
                 }
             }
         } else {
-            self.bip_path = restored_bip_path;
+            self.bip_path = loaded_bip_path;
         }
 
+        // Create the node client.
+        let client = self
+            .client_options
+            .clone()
+            .ok_or(crate::wallet::Error::MissingParameter("client_options"))?
+            .finish()
+            .await?;
+
+        match (self.address.as_ref(), self.bip_path.as_ref()) {
+            (Some(address), Some(bip_path)) => {
+                // verify that the address is derived from the provided bip path.
+                if let Some(backing_ed25519_address) = address.inner.backing_ed25519() {
+                    self.verify_ed25519_address(backing_ed25519_address, bip_path).await?;
```

Comment:
This seems wrong to me here, the secret manager shouldn't be needed at this point if the address is there, otherwise the secret manager needs to be unlocked/ready every time we build a wallet with existing storage?

Reply:
good point, we only need to check it hasn't been stored yet

Reply:
Changed